File: sort-algorithms/merge-sort/mergeSort.py
def mergeSort(arr: list):
  if len(arr) > 1: 
    middle = int(len(arr) / 2)
    left = arr[:middle]
    right = arr[middle:]
    
    arr1 = mergeSort(left)
    arr2 = mergeSort(right)
    
    arr = merge(arr1, arr2)
    
  return arr

def merge(left: list, right: list):
  resultArray = []
  leftIndex = rightIndex = 0
  
  while leftIndex < len(left) and rightIndex < len(right):
    if left[leftIndex] < right[rightIndex]:
      resultArray.append(left[leftIndex])
      leftIndex += 1
    else:
      resultArray.append(right[rightIndex])
      rightIndex += 1
    
  resultArray.extend(left[leftIndex:])
  resultArray.extend(right[rightIndex:])
  return resultArray


arr = [30,20,34,40,23,50,1,0]
print(mergeSort(arr))

File: sort-algorithms/merge-sort/mergeSort.js
function mergeSort(arr) {
  if(arr.length === 1) return arr;

  const middle = Math.floor(arr.length / 2);
  const left =  arr.slice(0, middle);
  const right = arr.slice(middle, arr.length);

  const arr1 = mergeSort(left);
  const arr2 = mergeSort(right);

  arr = merge(arr1, arr2);
  
  return arr;
};

function merge(left, right) {
  let resultArray = [],
    leftIndex = 0,
    rightIndex = 0;

  while (leftIndex < left.length && rightIndex < right.length) {
    if (left[leftIndex] < right[rightIndex]) {
      resultArray.push(left[leftIndex]);
      leftIndex++; 
    } else {
      resultArray.push(right[rightIndex]);
      rightIndex++;
    }
  }

  return resultArray
    .concat(left.slice(leftIndex))
    .concat(right.slice(rightIndex));
}
